File: SS/Labs/Lab5/ll1_parser.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

typedef struct {
    char lhs;
    char rhs[20];
} Prod;

char nt[] = {'E', 'A', 'T', 'B', 'F'};      // A = E', B = T'
char t[]  = {'i', '+', '*', '(', ')', '$'}; // i = id
Prod p[] = {
    {'E', "TA"}, {'A', "+TA"}, {'A', "e"}, {'T', "FB"},
    {'B', "*FB"}, {'B', "e"}, {'F', "i"}, {'F', "(E)"}
};

#define NNT ((int)(sizeof(nt) / sizeof(nt[0])))
#define NTM ((int)(sizeof(t) / sizeof(t[0])))
#define NPR ((int)(sizeof(p) / sizeof(p[0])))
#define SZ 20
#define MAX 200

char first[NNT][SZ], follow[NNT][SZ], table[NNT][NTM][SZ];

int has(char s[], char c) {
    for (int i = 0; s[i]; i++) if (s[i] == c) return 1;
    return 0;
}

int add(char s[], char c) {
    if (!c || has(s, c)) return 0;
    int n = strlen(s);
    s[n] = c;
    s[n + 1] = '\0';
    return 1;
}

int merge(char dst[], char src[], char skip) {
    int changed = 0;
    for (int i = 0; src[i]; i++) if (src[i] != skip) changed |= add(dst, src[i]);
    return changed;
}

int idx(char arr[], int n, char c) {
    for (int i = 0; i < n; i++) if (arr[i] == c) return i;
    return -1;
}

int nt_idx(char c) { return idx(nt, NNT, c); }
int t_idx(char c) { return idx(t, NTM, c); }
int is_terminal(char c) { return t_idx(c) != -1; }

void first_of_str(char s[], char out[]) {
    out[0] = '\0';
    if (!s[0] || (s[0] == 'e' && !s[1])) { add(out, 'e'); return; }

    for (int i = 0; s[i]; i++) {
        if (is_terminal(s[i])) { add(out, s[i]); return; }
        int ni = nt_idx(s[i]);
        if (ni == -1) return;
        merge(out, first[ni], 'e');
        if (!has(first[ni], 'e')) return;
    }
    add(out, 'e');
}

void compute_first() {
    for (int i = 0; i < NNT; i++) first[i][0] = '\0';

    int changed = 1;
    while (changed) {
        changed = 0;
        for (int i = 0; i < NPR; i++) {
            int a = nt_idx(p[i].lhs);
            char f[SZ];
            first_of_str(p[i].rhs, f);
            changed |= merge(first[a], f, '\0');
        }
    }
}

void compute_follow() {
    for (int i = 0; i < NNT; i++) follow[i][0] = '\0';
    add(follow[nt_idx('E')], '$');

    int changed = 1;
    while (changed) {
        changed = 0;
        for (int i = 0; i < NPR; i++) {
            int a = nt_idx(p[i].lhs);
            for (int j = 0; p[i].rhs[j]; j++) {
                int b = nt_idx(p[i].rhs[j]);
                if (b == -1) continue;

                char fs[SZ];
                first_of_str(p[i].rhs + j + 1, fs);
                changed |= merge(follow[b], fs, 'e');
                if (has(fs, 'e')) changed |= merge(follow[b], follow[a], '\0');
            }
        }
    }
}

void build_table() {
    for (int i = 0; i < NNT; i++)
        for (int j = 0; j < NTM; j++) strcpy(table[i][j], "-");

    for (int i = 0; i < NPR; i++) {
        int a = nt_idx(p[i].lhs);
        char f[SZ];
        first_of_str(p[i].rhs, f);

        for (int j = 0; f[j]; j++) {
            if (f[j] == 'e') continue;
            int c = t_idx(f[j]);
            if (c != -1) strcpy(table[a][c], p[i].rhs);
        }

        if (has(f, 'e')) {
            for (int j = 0; follow[a][j]; j++) {
                int c = t_idx(follow[a][j]);
                if (c != -1) strcpy(table[a][c], "e");
            }
        }
    }
}

int normalize(char raw[], char out[]) {
    int i = 0, k = 0;
    while (raw[i] && raw[i] != '\n') {
        if (isspace((unsigned char)raw[i])) { i++; continue; }
        if (raw[i] == 'i' && raw[i + 1] == 'd') { out[k++] = 'i'; i += 2; continue; }
        if (raw[i] == '+' || raw[i] == '*' || raw[i] == '(' || raw[i] == ')') { out[k++] = raw[i++]; continue; }
        return 0;
    }
    out[k++] = '$';
    out[k] = '\0';
    return 1;
}

int parse(char input[]) {
    char st[MAX];
    int top = -1, ip = 0;
    st[++top] = '$';
    st[++top] = 'E';

    while (top >= 0) {
        char X = st[top], a = input[ip];
        if (X == '$' && a == '$') return 1;

        if (is_terminal(X)) {
            if (X != a) return 0;
            top--;
            ip++;
            continue;
        }

        int r = nt_idx(X), c = t_idx(a);
        if (r == -1 || c == -1 || strcmp(table[r][c], "-") == 0) return 0;

        top--;
        if (strcmp(table[r][c], "e") != 0)
            for (int i = (int)strlen(table[r][c]) - 1; i >= 0; i--) st[++top] = table[r][c][i];
    }
    return 0;
}

void print_table() {
    printf("\nLL(1) Parse Table (A=E', B=T'):\n\n      ");
    for (int i = 0; i < NTM; i++) printf("%-7c ", t[i]);
    printf("\n");
    for (int i = 0; i < NNT; i++) {
        printf("%c   ", nt[i]);
        for (int j = 0; j < NTM; j++) printf("%-7s ", table[i][j]);
        printf("\n");
    }
}

int main() {
    char raw[MAX], input[MAX];

    compute_first();
    compute_follow();
    build_table();
    print_table();

    printf("\nEnter input string (use id, +, *, (, )): ");
    fgets(raw, sizeof(raw), stdin);

    if (!normalize(raw, input)) {
        printf("Invalid token in input!\n");
        return 0;
    }

    printf(parse(input) ? "Accepted\n" : "Rejected\n");
    return 0;
}

File: SS/Labs/PythonLabs/Lab5/ll1_parser.py
# Lab 5 — LL(1) Parser
# Grammar:
#   E  -> T A      (A = E')
#   A  -> + T A | epsilon
#   T  -> F B      (B = T')
#   B  -> * F B | epsilon
#   F  -> id | ( E )
# Terminals: id(i), +, *, (, ), $

NT = ['E', 'A', 'T', 'B', 'F']   # A = E', B = T'
TM = ['i', '+', '*', '(', ')', '$']   # i = id

prods = [
    ('E', 'TA'), ('A', '+TA'), ('A', 'e'), ('T', 'FB'),
    ('B', '*FB'), ('B', 'e'), ('F', 'i'),  ('F', '(E)')
]

# ── Helpers ──────────────────────────────────────────────────────────────

def nt_idx(c):
    return NT.index(c) if c in NT else -1

def t_idx(c):
    return TM.index(c) if c in TM else -1

def is_terminal(c):
    return t_idx(c) != -1

# ── FIRST of a string ────────────────────────────────────────────────────

def first_of_str(s, first_sets):
    out = set()
    if not s or s == 'e':
        return {'e'}
    for sym in s:
        if is_terminal(sym):
            out.add(sym)
            return out
        ni = nt_idx(sym)
        if ni == -1:
            return out
        out |= first_sets[ni] - {'e'}
        if 'e' not in first_sets[ni]:
            return out
    out.add('e')
    return out

# ── Compute FIRST ────────────────────────────────────────────────────────

def compute_first():
    first = [set() for _ in NT]
    changed = True
    while changed:
        changed = False
        for lhs, rhs in prods:
            a = nt_idx(lhs)
            f = first_of_str(rhs, first)
            before = len(first[a])
            first[a] |= f
            if len(first[a]) != before:
                changed = True
    return first

# ── Compute FOLLOW ───────────────────────────────────────────────────────

def compute_follow(first):
    follow = [set() for _ in NT]
    follow[nt_idx('E')].add('$')
    changed = True
    while changed:
        changed = False
        for lhs, rhs in prods:
            a = nt_idx(lhs)
            for j, sym in enumerate(rhs):
                b = nt_idx(sym)
                if b == -1:
                    continue
                fs = first_of_str(rhs[j+1:], first) if j + 1 < len(rhs) else {'e'}
                before = len(follow[b])
                follow[b] |= fs - {'e'}
                if 'e' in fs:
                    follow[b] |= follow[a]
                if len(follow[b]) != before:
                    changed = True
    return follow

# ── Build LL(1) Table ────────────────────────────────────────────────────

def build_table(first, follow):
    table = [['-'] * len(TM) for _ in NT]
    for i, (lhs, rhs) in enumerate(prods):
        a = nt_idx(lhs)
        f = first_of_str(rhs, first)
        for sym in f:
            if sym == 'e':
                continue
            c = t_idx(sym)
            if c != -1:
                table[a][c] = rhs
        if 'e' in f:
            for sym in follow[a]:
                c = t_idx(sym)
                if c != -1:
                    table[a][c] = 'e'
    return table

# ── Normalize input ──────────────────────────────────────────────────────

def normalize(raw):
    out = []
    i = 0
    raw = raw.strip()
    while i < len(raw):
        if raw[i:i+2] == 'id':
            out.append('i')
            i += 2
        elif raw[i] in '+*()':
            out.append(raw[i])
            i += 1
        elif raw[i] == ' ':
            i += 1
        else:
            return None  # invalid token
    out.append('$')
    return out

# ── Parse ────────────────────────────────────────────────────────────────

def parse(table, inp):
    stack = ['$', 'E']
    ip = 0
    while stack:
        X = stack[-1]
        a = inp[ip]
        if X == '$' and a == '$':
            return True
        if is_terminal(X):
            if X != a:
                return False
            stack.pop()
            ip += 1
        else:
            r = nt_idx(X)
            c = t_idx(a)
            if r == -1 or c == -1 or table[r][c] == '-':
                return False
            stack.pop()
            rhs = table[r][c]
            if rhs != 'e':
                for ch in reversed(rhs):
                    stack.append(ch)
    return False

# ── Print Table ──────────────────────────────────────────────────────────

def print_table(table):
    print("\nLL(1) Parse Table (A=E', B=T'):\n")
    header = "      " + "".join(f"{t:<8}" for t in TM)
    print(header)
    for i, row in enumerate(table):
        print(f"{NT[i]:<4}  " + "".join(f"{cell:<8}" for cell in row))

# ── Main ─────────────────────────────────────────────────────────────────

def main():
    first  = compute_first()
    follow = compute_follow(first)
    table  = build_table(first, follow)
    print_table(table)

    raw = input("\nEnter input string (use id, +, *, (, )): ").strip()
    inp = normalize(raw)
    if inp is None:
        print("Invalid token in input!")
    else:
        print("Accepted" if parse(table, inp) else "Rejected")

if __name__ == "__main__":
    main()
